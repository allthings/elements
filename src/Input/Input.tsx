import React, {
  useEffect,
  useState,
  Fragment,
  AllHTMLAttributes,
  FunctionComponent,
  forwardRef,
  useCallback,
  useRef,
} from 'react'
import View from '../View'
import { css } from 'glamor'
import Relative from '../Relative'
import Text, { createTextStyles } from '../Text'
import Icon, { IconType } from '../Icon'
import Absolute from '../Absolute'
import { useCombinedRefs } from '../utils/hooks/useCombinedRefs'

const styles = {
  input: (showLabel: boolean, translated: boolean, paddingRight: boolean) =>
    css(createTextStyles({ size: 'm' }), {
      boxSizing: 'border-box',
      height: 50,
      width: '100%',
      padding: '0 15px',
      paddingLeft: translated && 40,
      paddingTop: showLabel ? 10 : 0,
      paddingRight: paddingRight ? 50 : 15,
      transition: 'padding-top .225s ease-out',
      border: 0,
      '&:-webkit-autofill ~ .label': {
        opacity: '1 !important',
        top: '8px !important',
      },
      '&:-webkit-autofill ~ .checkmark': {
        opacity: '1 !important',
        top: 8,
      },
      '&:-webkit-autofill': {
        paddingTop: '10px !important',
      },
    }),
  area: (lines: number, showLabel: boolean) =>
    css(createTextStyles({ size: 'm' }), {
      boxSizing: 'border-box',
      transition: 'padding-top .225s ease-out',
      height: `calc(30px*${lines})`,
      width: '100%',
      padding: '10px 15px',
      paddingTop: showLabel ? 20 : 10,
      fontSize: '14px',
      border: 0,
      ':invalid:focus': {
        color: 'red',
      },
    }),
  arrow: css({
    position: 'absolute',
    bottom: '-10px',
    left: '50%',
    marginLeft: '-10px',
    display: 'block',
    width: 0,
    height: 0,
    borderLeft: '10px solid transparent',
    borderRight: '10px solid transparent',
    borderTop: '10px solid #c1392b',
  }),
  checkmark: (show: boolean) =>
    css({
      position: 'absolute',
      top: 16,
      right: 15,
      pointerEvents: 'none',
      transition: 'opacity .225s',
      opacity: show ? 1 : 0,
    }),
  required: css({
    position: 'absolute',
    right: 10,
  }),
  label: (translated: boolean) =>
    css({
      position: 'absolute',
      left: translated ? 40 : 15,
      fontSize: 10,
      opacity: 0,
      transition: 'all .225s ease-out',
    }),
  placeholder: css({
    position: 'absolute',
    bottom: 2,
    right: 15,
  }),
}

const validityStates = [
  'badInput',
  'customError',
  'patternMismatch',
  'rangeOverflow',
  'rangeUnderflow',
  'stepMismatch',
  'tooLong',
  'tooShort',
  'typeMismatch',
  'valueMissing',
] as const

type validityStateType = typeof validityStates[number]

export interface IInputProps
  extends IValidityStates,
    AllHTMLAttributes<HTMLInputElement | HTMLTextAreaElement> {
  /** The default value to put into the component, without making it controlled */
  readonly defaultValue?: string | string[]
  /** Indicates that this field is required */
  readonly required?: boolean
  /** Icon shown on the left of the input field (See `Icon` component for all possible values) **/
  readonly icon?: IconType
  /** The name of this input field */
  readonly name?: string
  /** The label of the input */
  readonly label?: string
  /** Type, can be: 'tel', 'number', 'text', 'url', 'email' */
  readonly type?:
    | 'tel'
    | 'number'
    | 'text'
    | 'url'
    | 'email'
    | 'password'
    | 'date'
    | 'datetime-local'
    | string
  /** Called, when the user changes something */
  readonly onChange?: any
  /** The value, makes this component a controlled component */
  readonly value?: string | string[] | number
  /** Can only be used with type=text. Increase to enable multi-line input */
  readonly lines?: number
  /** Used when there is an icon to the right of input field */
  readonly hasRightIcon?: boolean
  /** Regular expression to validate against */
  readonly pattern?: string
  /** Min number of characters that must be provided */
  readonly minLength?: number
  /** Max number of characters that can be provided */
  readonly maxLength?: number
  /** Called with the input field a reference */
  readonly onInputRef?: any
  readonly readOnly?: boolean
  readonly disabled?: boolean
  readonly forwardedRef?: any
}

interface IValidityStates {
  /** Error message for bad input */
  readonly badInput?: string
  /** Error message for customError */
  readonly customError?: string
  /** Error message for patternMismatch */
  readonly patternMismatch?: string
  /** Error message for rangeOverflow */
  readonly rangeOverflow?: string
  /** Error message for rangeUnderflow */
  readonly rangeUnderflow?: string
  /** Error message for stepMismatch */
  readonly stepMismatch?: string
  /** Error message for tooLong */
  readonly tooLong?: string
  /** Error message for tooShort */
  readonly tooShort?: string
  /** Error message for typeMismatch */
  readonly typeMismatch?: string
  /** Error message for valueMissing */
  readonly valueMissing?: string
}

const setValidity = (
  target: HTMLInputElement | HTMLTextAreaElement,
  customValidityStates: IValidityStates,
) => {
  let customValidity = ''
  validityStates.forEach((state: validityStateType) => {
    if (
      customValidity === '' &&
      customValidityStates[state] &&
      target.validity[state]
    ) {
      customValidity = customValidityStates[state] as string
    }
  })

  target && target.setCustomValidity && target.setCustomValidity(customValidity)
}

/**
 * TextInputs are used to allow users to enter information like names, numbers, urls, email addresses or passwords.
 *
 * ```example
 * <TextInput name="email" type="email" placeholder="Your email" required />
 * <TextInput name="inquiry" lines={5} placeholder="Your question" maxLength={255} minLength={50} />
 * ```
 */

const Input: FunctionComponent<IInputProps> = ({
  required = false,
  forwardedRef,
  lines = 1,
  label,
  pattern,
  badInput,
  customError,
  hasRightIcon,
  patternMismatch,
  rangeOverflow,
  rangeUnderflow,
  stepMismatch,
  tooLong,
  type = 'text',
  tooShort,
  typeMismatch,
  valueMissing,
  icon,
  ...props
}) => {
  const internalRef = useRef<any>(null)
  const isTextArea = lines !== 1
  const [value, setValue] = useState('')
  const [length, setLength] = useState(
    (typeof props.value === 'string' && props.value.length) || 0,
  )
  const currentValue = (props.value || value) as string
  const labelVisible = currentValue.length > 0
  const showLabel = !!(label && currentValue.length > 0)

  const isCheckmarkActive = Boolean(
    (pattern || props.minLength || props.maxLength || required) &&
      internalRef &&
      internalRef.current &&
      internalRef.current.validity &&
      internalRef.current.validity.valid,
  )

  const customValidity = {
    rangeOverflow,
    rangeUnderflow,
    stepMismatch,
    tooLong,
    tooShort,
    typeMismatch,
    valueMissing,
  }

  const handleChange = useCallback(
    (
      e:
        | React.ChangeEvent<HTMLInputElement>
        | React.ChangeEvent<HTMLTextAreaElement>,
    ) => {
      setValidity(e.target, customValidity)
      setLength(e.target.value.length)
      setValue(e.target.value)
      props.onChange && props.onChange(e)
    },
    [customValidity, props],
  )

  useEffect(() => {
    const input = internalRef.current
    if (input) {
      setLength(input.value && input.value.length ? input.value.length : 0)
      setValidity(input, customValidity)
    }
  }, [isTextArea, customValidity])

  const combinedRef = useCombinedRefs(internalRef, forwardedRef)

  return (
    <Relative style={{ width: '100%' }}>
      {!isTextArea ? (
        <Fragment>
          {icon && (
            <Absolute
              {...css({ pointerEvents: 'none' })}
              top={0}
              bottom={0}
              left={15}
              alignV="center"
              direction="row"
            >
              <Icon color="secondaryText" name={icon} size={16} />
            </Absolute>
          )}
          <input
            type={type}
            ref={combinedRef}
            {...styles.input(
              showLabel,
              !!icon,
              Boolean(isCheckmarkActive || hasRightIcon),
            )}
            required={required}
            aria-required={required}
            {...props}
            onChange={handleChange}
            pattern={pattern}
          />
        </Fragment>
      ) : (
        <textarea
          ref={combinedRef}
          {...styles.area(lines, showLabel)}
          required={required}
          {...props}
          onChange={handleChange}
        />
      )}
      {label && (
        <View
          className="label"
          {...styles.label(!!icon && lines === 1)}
          style={{
            opacity: labelVisible ? 1 : 0,
            top: labelVisible ? 8 : 12,
          }}
        >
          <Text color="secondaryText" size="xs">
            {label} {required && '*'}
          </Text>
        </View>
      )}

      <View className="checkmark" {...styles.checkmark(isCheckmarkActive)}>
        <Icon name="check-filled" size="xs" color="lightGrey" />
      </View>

      {props.maxLength && (
        <View {...styles.placeholder}>
          <Text color="secondaryText" size="s">
            {length}/{props.maxLength}
          </Text>
        </View>
      )}
    </Relative>
  )
}

export default forwardRef<HTMLInputElement, IInputProps>((props, ref) => (
  <Input {...props} forwardedRef={ref} />
))
