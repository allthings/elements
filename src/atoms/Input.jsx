import React, { Fragment } from 'react'
import View from '../atoms/View'
import { css } from 'glamor'
import PropTypes from 'prop-types'
import Relative from '../atoms/Relative'
import Text, { createTextStyles } from '../atoms/Text'
import Icon, { Icons as AvailableIcons } from '../atoms/Icon'
import Absolute from './Absolute'

const styles = {
  input: (showLabel, translated, paddingRight) =>
    css(createTextStyles({ size: 'm' }), {
      boxSizing: 'border-box',
      height: 50,
      width: '100%',
      padding: '0 15px',
      paddingLeft: translated && 40,
      paddingTop: showLabel ? 10 : 0,
      paddingRight: paddingRight ? 50 : 15,
      transition: 'padding-top .225s ease-out',
      border: 0,
      '&:-webkit-autofill ~ .label': {
        opacity: '1 !important',
        top: '8px !important',
      },
      '&:-webkit-autofill ~ .checkmark': {
        opacity: '1 !important',
        top: 8,
      },
      '&:-webkit-autofill': {
        paddingTop: '10px !important',
      },
    }),
  area: (lines, showLabel) =>
    css(createTextStyles({ size: 'm' }), {
      boxSizing: 'border-box',
      transition: 'padding-top .225s ease-out',
      height: `calc(30px*${lines})`,
      width: '100%',
      padding: '10px 15px',
      paddingTop: showLabel ? 20 : 10,
      fontSize: '14px',
      border: 0,
      ':invalid:focus': {
        color: 'red',
      },
    }),
  arrow: css({
    position: 'absolute',
    bottom: '-10px',
    left: '50%',
    marginLeft: '-10px',
    display: 'block',
    width: 0,
    height: 0,
    borderLeft: '10px solid transparent',
    borderRight: '10px solid transparent',
    borderTop: '10px solid #c1392b',
  }),
  checkmark: show =>
    css({
      position: 'absolute',
      top: 16,
      right: 15,
      pointerEvents: 'none',
      transition: 'opacity .225s',
      opacity: show ? 1 : 0,
    }),
  required: css({
    position: 'absolute',
    right: 10,
  }),
  label: translated =>
    css({
      position: 'absolute',
      left: translated ? 40 : 15,
      fontSize: 10,
      opacity: 0,
      transition: 'all .225s ease-out',
    }),
  placeholder: css({
    position: 'absolute',
    bottom: 2,
    right: 15,
  }),
}

const validityStates = [
  'badInput',
  'customError',
  'patternMismatch',
  'rangeOverflow',
  'rangeUnderflow',
  'stepMismatch',
  'tooLong',
  'tooShort',
  'typeMismatch',
  'valueMissing',
]

/**
 * TextInputs are used to allow users to enter information like names, numbers, urls, email addresses or passwords.
 *
 * ```example
 * <TextInput name="email" type="email" placeholder="Your email" required />
 * <TextInput name="inquiry" lines={5} placeholder="Your question" maxLength={255} minLength={50} />
 * ```
 */
class Input extends React.Component {
  static propTypes = {
    /** The default value to put into the component, without making it controlled */
    defaultValue: PropTypes.string,
    /** Indicates that this field is required */
    required: PropTypes.bool,
    /** Icon shown on the left of the input field (See `Icon` component for all possible values) **/
    icon: PropTypes.oneOf(AvailableIcons),
    /** The name of this input field */
    name: PropTypes.string.isRequired,
    /** The label of the input */
    label: PropTypes.string,
    /** Type, can be: 'tel', 'number', 'text', 'url', 'email' */
    type: PropTypes.oneOf([
      'tel',
      'number',
      'text',
      'url',
      'email',
      'password',
      'date',
      'datetime-local',
    ]),
    /** Called, when the user changes something */
    onChange: PropTypes.func,
    /** The value, makes this component a controlled component */
    value: PropTypes.string,
    /** Can only be used with type=text. Increase to enable multi-line input */
    lines: PropTypes.number,
    /** Used when there is an icon to the right of input field */
    paddingRight: PropTypes.bool,
    /** Regular expression to validate against */
    pattern: PropTypes.string,
    /** Min number of characters that must be provided */
    minLength: PropTypes.number,
    /** Max number of characters that can be provided */
    maxLength: PropTypes.number,
    /** Called with the input field a reference */
    onInputRef: PropTypes.func,
    /** Error message for bad input */
    badInput: PropTypes.string,
    /** Error message for customError */
    customError: PropTypes.string,
    /** Error message for patternMismatch */
    patternMismatch: PropTypes.string,
    /** Error message for rangeOverflow */
    rangeOverflow: PropTypes.string,
    /** Error message for rangeUnderflow */
    rangeUnderflow: PropTypes.string,
    /** Error message for stepMismatch */
    stepMismatch: PropTypes.string,
    /** Error message for tooLong */
    tooLong: PropTypes.string,
    /** Error message for tooShort */
    tooShort: PropTypes.string,
    /** Error message for typeMismatch */
    typeMismatch: PropTypes.string,
    /** Error message for valueMissing */
    valueMissing: PropTypes.string,
  }

  state = {
    value: '',
    visible: true,
    length: (this.props.value && this.props.value.length) || 0,
  }

  static defaultProps = {
    required: false,
    lines: 1,
    onInputRef: _ => _,
    type: 'text',
  }

  input = React.createRef()

  setValidity = target => {
    let customValidity = ''
    validityStates.forEach(state => {
      if (
        customValidity === '' &&
        this.props[state] &&
        target.validity[state]
      ) {
        customValidity = this.props[state]
      }
    })

    target &&
      target.setCustomValidity &&
      target.setCustomValidity(customValidity)
  }

  componentDidMount() {
    const input = this.input.current
    if (input) {
      this.setState({ length: input.value && input.value.length })
      this.props.onInputRef(input)
      this.setValidity(input)
    }
  }

  handleChange = e => {
    this.setValidity(e.target)
    this.setState({
      value: e.target.value,
      length: e.target.value.length,
    })
    this.props.onChange && this.props.onChange(e)
  }

  render() {
    const {
      required,
      onInputRef,
      lines,
      label,
      pattern,
      badInput,
      customError,
      paddingRight,
      patternMismatch,
      rangeOverflow,
      rangeUnderflow,
      stepMismatch,
      tooLong,
      tooShort,
      typeMismatch,
      valueMissing,
      icon,
      ...props
    } = this.props
    const currentValue = this.props.value || this.state.value
    const labelVisible = currentValue.length > 0
    const showLabel = label && currentValue.length > 0

    const isCheckmarkActive =
      (pattern || props.minLength || props.maxLength || required) &&
      this.input.current &&
      this.input.current.validity &&
      this.input.current.validity.valid

    return (
      <Relative style={{ width: '100%' }}>
        {lines === 1 ? (
          <Fragment>
            {icon && (
              <Absolute
                {...css({ pointerEvents: 'none' })}
                top={0}
                bottom={0}
                left={15}
                alignV="center"
                direction="row"
              >
                <Icon color="secondaryText" name={icon} size={16} />
              </Absolute>
            )}
            <input
              ref={this.input}
              {...styles.input(
                showLabel,
                !!icon,
                isCheckmarkActive || paddingRight
              )}
              required={required}
              aria-required={required}
              {...props}
              onChange={this.handleChange}
              pattern={pattern}
            />
          </Fragment>
        ) : (
          <textarea
            ref={this.input}
            {...styles.area(lines, showLabel)}
            required={required}
            {...props}
            onChange={this.handleChange}
          />
        )}
        {label && (
          <View
            className="label"
            {...styles.label(!!icon && lines === 1)}
            style={{
              opacity: labelVisible ? 1 : 0,
              top: labelVisible ? 8 : 12,
            }}
          >
            <Text color="secondaryText" size="xs">
              {label} {required && '*'}
            </Text>
          </View>
        )}

        <View className="checkmark" {...styles.checkmark(isCheckmarkActive)}>
          <Icon name="check-filled" size="xs" color="lightGrey" />
        </View>

        {props.maxLength && (
          <View {...styles.placeholder}>
            <Text color="secondaryText" size="s">
              {this.state.length}/{props.maxLength}
            </Text>
          </View>
        )}
      </Relative>
    )
  }
}

export default Input
