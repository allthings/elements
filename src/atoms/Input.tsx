import React, { useEffect, useState, Fragment } from 'react'
import View from '../atoms/View'
import { css } from 'glamor'
import Relative from '../atoms/Relative'
import Text, { createTextStyles } from '../atoms/Text'
import Icon, { IconType } from '../atoms/Icon'
import Absolute from './Absolute'

const styles = {
  input: (showLabel: boolean, translated: boolean, paddingRight: boolean) =>
    css(createTextStyles({ size: 'm' }), {
      boxSizing: 'border-box',
      height: 50,
      width: '100%',
      padding: '0 15px',
      paddingLeft: translated && 40,
      paddingTop: showLabel ? 10 : 0,
      paddingRight: paddingRight ? 50 : 15,
      transition: 'padding-top .225s ease-out',
      border: 0,
      '&:-webkit-autofill ~ .label': {
        opacity: '1 !important',
        top: '8px !important',
      },
      '&:-webkit-autofill ~ .checkmark': {
        opacity: '1 !important',
        top: 8,
      },
      '&:-webkit-autofill': {
        paddingTop: '10px !important',
      },
    }),
  area: (lines: number, showLabel: boolean) =>
    css(createTextStyles({ size: 'm' }), {
      boxSizing: 'border-box',
      transition: 'padding-top .225s ease-out',
      height: `calc(30px*${lines})`,
      width: '100%',
      padding: '10px 15px',
      paddingTop: showLabel ? 20 : 10,
      fontSize: '14px',
      border: 0,
      ':invalid:focus': {
        color: 'red',
      },
    }),
  arrow: css({
    position: 'absolute',
    bottom: '-10px',
    left: '50%',
    marginLeft: '-10px',
    display: 'block',
    width: 0,
    height: 0,
    borderLeft: '10px solid transparent',
    borderRight: '10px solid transparent',
    borderTop: '10px solid #c1392b',
  }),
  checkmark: (show: boolean) =>
    css({
      position: 'absolute',
      top: 16,
      right: 15,
      pointerEvents: 'none',
      transition: 'opacity .225s',
      opacity: show ? 1 : 0,
    }),
  required: css({
    position: 'absolute',
    right: 10,
  }),
  label: (translated: boolean) =>
    css({
      position: 'absolute',
      left: translated ? 40 : 15,
      fontSize: 10,
      opacity: 0,
      transition: 'all .225s ease-out',
    }),
  placeholder: css({
    position: 'absolute',
    bottom: 2,
    right: 15,
  }),
}

const validityStates = [
  'badInput',
  'customError',
  'patternMismatch',
  'rangeOverflow',
  'rangeUnderflow',
  'stepMismatch',
  'tooLong',
  'tooShort',
  'typeMismatch',
  'valueMissing',
]

type validityStateType =
  | 'badInput'
  | 'customError'
  | 'patternMismatch'
  | 'rangeOverflow'
  | 'rangeUnderflow'
  | 'stepMismatch'
  | 'tooLong'
  | 'tooShort'
  | 'typeMismatch'
  | 'valueMissing'

interface IProps extends IValidityStates {
  /** The default value to put into the component, without making it controlled */
  defaultValue?: string
  /** Indicates that this field is required */
  required?: boolean
  /** Icon shown on the left of the input field (See `Icon` component for all possible values) **/
  icon?: IconType
  /** The name of this input field */
  name?: string
  /** The label of the input */
  label?: string
  /** Type, can be: 'tel', 'number', 'text', 'url', 'email' */
  type?:
    | 'tel'
    | 'number'
    | 'text'
    | 'url'
    | 'email'
    | 'password'
    | 'date'
    | 'datetime-local'
  /** Called, when the user changes something */
  onChange?: any
  /** The value, makes this component a controlled component */
  value?: string
  /** Can only be used with type=text. Increase to enable multi-line input */
  lines?: number
  /** Used when there is an icon to the right of input field */
  hasRightIcon?: boolean
  /** Regular expression to validate against */
  pattern?: string
  /** Min number of characters that must be provided */
  minLength?: number
  /** Max number of characters that can be provided */
  maxLength?: number
  /** Called with the input field a reference */
  onInputRef?: any
  readOnly?: boolean
  disabled?: boolean
}

interface IValidityStates {
  /** Error message for bad input */
  badInput?: string
  /** Error message for customError */
  customError?: string
  /** Error message for patternMismatch */
  patternMismatch?: string
  /** Error message for rangeOverflow */
  rangeOverflow?: string
  /** Error message for rangeUnderflow */
  rangeUnderflow?: string
  /** Error message for stepMismatch */
  stepMismatch?: string
  /** Error message for tooLong */
  tooLong?: string
  /** Error message for tooShort */
  tooShort?: string
  /** Error message for typeMismatch */
  typeMismatch?: string
  /** Error message for valueMissing */
  valueMissing?: string
}

const setValidity = (
  target: HTMLInputElement | HTMLTextAreaElement,
  customValidityStates: IValidityStates
) => {
  let customValidity = ''
  validityStates.forEach((state: validityStateType) => {
    if (
      customValidity === '' &&
      customValidityStates[state] &&
      target.validity[state]
    ) {
      customValidity = customValidityStates[state] as string
    }
  })

  target && target.setCustomValidity && target.setCustomValidity(customValidity)
}

/**
 * TextInputs are used to allow users to enter information like names, numbers, urls, email addresses or passwords.
 *
 * ```example
 * <TextInput name="email" type="email" placeholder="Your email" required />
 * <TextInput name="inquiry" lines={5} placeholder="Your question" maxLength={255} minLength={50} />
 * ```
 */

const Input = ({
  required = false,
  onInputRef,
  lines = 1,
  label,
  pattern,
  badInput,
  customError,
  hasRightIcon,
  patternMismatch,
  rangeOverflow,
  rangeUnderflow,
  stepMismatch,
  tooLong,
  type = 'text',
  tooShort,
  typeMismatch,
  valueMissing,
  icon,
  ...props
}: IProps & React.HTMLAttributes<HTMLElement>) => {
  const isTextArea = lines !== 1
  const [value, setValue] = useState('')
  const [length, setLength] = useState((props.value && props.value.length) || 0)
  const currentValue = props.value || value
  const labelVisible = currentValue.length > 0
  const showLabel = !!(label && currentValue.length > 0)
  const inputRef = React.createRef<HTMLInputElement>()
  const textareaRef = React.createRef<HTMLTextAreaElement>()
  const isCheckmarkActive =
    (pattern || props.minLength || props.maxLength || required) &&
    inputRef.current &&
    inputRef.current.validity &&
    inputRef.current.validity.valid

  const customValidity = {
    rangeOverflow,
    rangeUnderflow,
    stepMismatch,
    tooLong,
    tooShort,
    typeMismatch,
    valueMissing,
  }

  const handleChange = (
    e:
      | React.ChangeEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLTextAreaElement>
  ) => {
    setValidity(e.target, customValidity)
    setLength(e.target.value.length)
    setValue(e.target.value)
    props.onChange && props.onChange(e)
  }

  useEffect(() => {
    const input = isTextArea ? textareaRef.current : inputRef.current
    if (input) {
      setLength(input.value && input.value.length ? input.value.length : 0)
      onInputRef && onInputRef(input)
      setValidity(input, customValidity)
    }
  })

  return (
    <Relative style={{ width: '100%' }}>
      {!isTextArea ? (
        <Fragment>
          {icon && (
            <Absolute
              {...css({ pointerEvents: 'none' })}
              top={0}
              bottom={0}
              left={15}
              alignV="center"
              direction="row"
            >
              <Icon color="secondaryText" name={icon} size={16} />
            </Absolute>
          )}
          <input
            type={type}
            ref={inputRef}
            {...styles.input(
              showLabel,
              !!icon,
              Boolean(isCheckmarkActive || hasRightIcon)
            )}
            required={required}
            aria-required={required}
            {...props}
            onChange={handleChange}
            pattern={pattern}
          />
        </Fragment>
      ) : (
        <textarea
          ref={textareaRef}
          {...styles.area(lines, showLabel)}
          required={required}
          {...props}
          onChange={handleChange}
        />
      )}
      {label && (
        <View
          className="label"
          {...styles.label(!!icon && lines === 1)}
          style={{
            opacity: labelVisible ? 1 : 0,
            top: labelVisible ? 8 : 12,
          }}
        >
          <Text color="secondaryText" size="xs">
            {label} {required && '*'}
          </Text>
        </View>
      )}

      <View className="checkmark" {...styles.checkmark(!!isCheckmarkActive)}>
        <Icon name="check-filled" size="xs" color="lightGrey" />
      </View>

      {props.maxLength && (
        <View {...styles.placeholder}>
          <Text color="secondaryText" size="s">
            {length}/{props.maxLength}
          </Text>
        </View>
      )}
    </Relative>
  )
}

export default Input
