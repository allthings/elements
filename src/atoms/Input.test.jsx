import React from 'react'
import renderer from 'react-test-renderer'
import Input from './Input'

import genProps from 'react-generate-props'

describe('Input component', () => {
  test('should render with all props', () => {
    const props = genProps(Input, { optional: true })
    const tree = renderer.create(<Input {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should render with required props only', () => {
    const props = genProps(Input)
    const tree = renderer.create(<Input {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
