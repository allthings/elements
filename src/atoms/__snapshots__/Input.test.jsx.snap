// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should render with all props 1`] = `
.css-0,
[data-css-0] {
  position: absolute;
  bottom: 2px;
  right: 15px;
}

.css-1,
[data-css-1] {
  pointer-events: none;
}

.css-2,
[data-css-2] {
  display: inline;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 13px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  box-sizing: border-box;
  height: 50px;
  width: 100%;
  padding: 0 15px;
  padding-left: 40px;
  padding-top: 10px;
  padding-right: 50px;
  transition: padding-top .225s ease-out;
  border: 0;
  -webkit-transition: padding-top .225s ease-out;
  -moz-transition: padding-top .225s ease-out;
}

.css-2:required,
[data-css-2]:required {
  box-shadow: none;
}

.css-2:-webkit-autofill ~ .label,
[data-css-2]:-webkit-autofill ~ .label {
  opacity: 1 !important;
  top: 8px !important;
}

.css-2:-webkit-autofill ~ .checkmark,
[data-css-2]:-webkit-autofill ~ .checkmark {
  opacity: 1 !important;
  top: 8px;
}

.css-2:-webkit-autofill,
[data-css-2]:-webkit-autofill {
  padding-top: 10px !important;
}

.css-2::-ms-clear,
[data-css-2]::-ms-clear {
  display: none;
}

.css-6,
[data-css-6] {
  position: absolute;
  left: 40px;
  font-size: 10px;
  opacity: 0;
  transition: all .225s ease-out;
  -webkit-transition: all .225s ease-out;
  -moz-transition: all .225s ease-out;
}

.css-7,
[data-css-7] {
  position: absolute;
  top: 16px;
  right: 15px;
  pointer-events: none;
  transition: opacity .225s;
  opacity: 0;
  -webkit-transition: opacity .225s;
  -moz-transition: opacity .225s;
}

.css-8,
[data-css-8] {
  position: relative;
}

.css-9,
[data-css-9] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-10,
[data-css-10] {
  position: absolute;
  top: 0;
  left: 15px;
  bottom: 0;
}

.css-11,
[data-css-11] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 0 0 auto;
}

.css-12,
[data-css-12] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 10px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #626262;
}

.css-13,
[data-css-13] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 12px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #626262;
}

<div
  data-css-8=""
  data-css-9=""
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    data-css-1=""
    data-css-11=""
    data-css-10=""
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
      data-css-9=""
      style={
        Object {
          "fill": false,
          "height": 16,
          "stroke": "#626262",
          "width": 16,
        }
      }
    />
  </div>
  <input
    aria-required={true}
    data-css-2=""
    defaultValue="defaultValue"
    maxLength={1}
    minLength={1}
    name="name"
    onChange={[Function]}
    pattern={[Function]}
    required={true}
    type="tel"
    value="value"
  />
  <div
    className="label"
    data-css-6=""
    data-css-9=""
    style={
      Object {
        "opacity": 1,
        "top": 8,
      }
    }
  >
    <div
      data-css-9=""
      data-css-12=""
    >
      label
       
      *
    </div>
  </div>
  <div
    className="checkmark"
    data-css-7=""
    data-css-9=""
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
      data-css-9=""
      style={
        Object {
          "fill": "lightGrey",
          "height": 16,
          "stroke": false,
          "width": 16,
        }
      }
    />
  </div>
  <div
    data-css-9=""
    data-css-0=""
  >
    <div
      data-css-13=""
      data-css-9=""
    >
      5
      /
      1
    </div>
  </div>
</div>
`;

exports[`Input component should render with required props only 1`] = `
.css-0,
[data-css-0] {
  position: absolute;
  top: 16px;
  right: 15px;
  pointer-events: none;
  transition: opacity .225s;
  opacity: 0;
  -webkit-transition: opacity .225s;
  -moz-transition: opacity .225s;
}

.css-1,
[data-css-1] {
  position: relative;
}

.css-2,
[data-css-2] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-3,
[data-css-3] {
  display: inline;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 13px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  box-sizing: border-box;
  height: 50px;
  width: 100%;
  padding: 0 15px;
  padding-top: 0;
  padding-right: 15px;
  transition: padding-top .225s ease-out;
  border: 0;
  -webkit-transition: padding-top .225s ease-out;
  -moz-transition: padding-top .225s ease-out;
}

.css-3:required,
[data-css-3]:required {
  box-shadow: none;
}

.css-3:-webkit-autofill ~ .label,
[data-css-3]:-webkit-autofill ~ .label {
  opacity: 1 !important;
  top: 8px !important;
}

.css-3:-webkit-autofill ~ .checkmark,
[data-css-3]:-webkit-autofill ~ .checkmark {
  opacity: 1 !important;
  top: 8px;
}

.css-3:-webkit-autofill,
[data-css-3]:-webkit-autofill {
  padding-top: 10px !important;
}

.css-3::-ms-clear,
[data-css-3]::-ms-clear {
  display: none;
}

<div
  data-css-1=""
  data-css-2=""
  style={
    Object {
      "width": "100%",
    }
  }
>
  <input
    aria-required={false}
    data-css-3=""
    name="name"
    onChange={[Function]}
    required={false}
    type="text"
  />
  <div
    className="checkmark"
    data-css-0=""
    data-css-2=""
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
      data-css-2=""
      style={
        Object {
          "fill": "lightGrey",
          "height": 16,
          "stroke": false,
          "width": 16,
        }
      }
    />
  </div>
</div>
`;
